java/src/main/java/io/cucumber/java/Step.java

```diff
@@ -0,0 +1,68 @@
+package io.cucumber.java;
+
+import io.cucumber.core.backend.TestCaseState;
+import io.cucumber.plugin.event.HookTestStep;
+import io.cucumber.plugin.event.Location;
+import io.cucumber.plugin.event.PickleStepTestStep;
+import io.cucumber.plugin.event.StepArgument;
+import io.cucumber.plugin.event.TestStep;
+import org.apiguardian.api.API;
+
+import java.util.Optional;
+
+/**
+ * BeforeStep or AfterStep Hooks that declare a parameter of this type will
+ * receive an instance of this class.
+ */
+@API(status = API.Status.EXPERIMENTAL)
+public class Step implements io.cucumber.plugin.event.Step {
+
+    final TestCaseState testCaseState;
+
+    final io.cucumber.plugin.event.Step delegate;
+
+    public Step(TestCaseState state) {
+        this.testCaseState = state;
+        Optional<TestStep> currentTestStep = state.getCurrentTestStep();
+        if (!currentTestStep.isPresent()) {
+            throw new IllegalStateException("No current TestStep was found in TestCaseState");
+        }
+        if (!(currentTestStep.get() instanceof HookTestStep)) {
+            throw new IllegalStateException("Current TestStep is not a HookTestStep");
+        }
+        TestStep relatedTestStep = ((HookTestStep) currentTestStep.get()).getRelatedTestStep();
+        if (null == relatedTestStep) {
+            throw new IllegalStateException("No related TestStep for current HookTestStep was found");
+        }
+        if (!(relatedTestStep instanceof PickleStepTestStep)) {
+            throw new IllegalStateException("Related TestStep is not a PickleStepTestStep");
```

Comment:
Given that most of these conditions this method checks for can only happen in case of a bug in Cucumber, consider using `Optional.filter`, and friends to make this more readable.